secrets/provider.go

```diff
@@ -0,0 +1,92 @@
+package secrets
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	corev2 "github.com/sensu/sensu-go/api/core/v2"
+)
+
+// A Provider gets all secrets from a secrets provider.
+type Provider interface {
+	corev2.Resource
+	// Get gets the value of the secret associated with the Sensu resource name.
+	Get(name string) (secret string, err error)
+}
+
+// ProviderManager manages the list of secrets providers.
+type ProviderManager struct {
+	mu        sync.RWMutex
+	providers map[string]Provider
+}
+
+// AddProvider adds a provider to the list of configured providers.
+func (m *ProviderManager) AddProvider(provider Provider) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Make sure the providers map is not nil
+	if m.providers == nil {
+		m.providers = map[string]Provider{}
+	}
+
+	m.providers[provider.GetObjectMeta().Name] = provider
+}
+
+// Providers returns the configured providers.
+func (m *ProviderManager) Providers() map[string]Provider {
+	m.mu.RLock()
```

Comment:
I'm playing with this PR and I just hit a panic here:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x4e246b9]

goroutine 485 [running]:
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:48
github.com/sensu/sensu-go/secrets.(*ProviderManager).Providers(0x0, 0x0)
	/Users/splourde/src/sensu-go/secrets/provider.go:39 +0x49
github.com/sensu/sensu-go/secrets.(*ProviderManager).SubSecrets(0x0, 0xc000f60090, 0x9, 0x6, 0xc000a0f558, 0x0, 0x0, 0x0)
	/Users/splourde/src/sensu-go/secrets/provider.go:74 +0x139
github.com/sensu/sensu-go/backend/schedulerd.buildRequest(0xc000ad01c0, 0x55cadc0, 0xc0004a5300, 0x0, 0x0, 0x0, 0x0)
	/Users/splourde/src/sensu-go/backend/schedulerd/executor.go:342 +0x2b0
```

I'd assume it's simply because the mutex `m.mu` is never initialized!

Reply:
Do you recall how you produced this panic?

Reply:
I literally did nothing, I just have to wait ~10 seconds. However I do have an interval check created, which I would assume is required so the scheduler tries to schedule it.

Reply:
I ran several types of checks before I submitted this PR, and again just now, and I still can't reproduce the panic. The provider manager mirrors that for [auth providers](https://github.com/sensu/sensu-go/blob/517d98798d5972c8d3848c9e4cee415fb652079c/backend/authentication/authenticator.go#L68-L113), and I don't see any initialization of the mutex anywhere. Perhaps I'm missing something..

Reply:
In any case, I will make sure it's initialized and we can see if that solves your issue?

Reply:
Okay added an initializing function, let me know if that doesn't resolve the panic! @palourde